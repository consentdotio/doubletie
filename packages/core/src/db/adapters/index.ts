/**
 * Database adapters registry
 *
 * This file contains the adapter registry for database adapters.
 * Each database type (SQLite, MySQL, PostgreSQL) has its own adapter implementation.
 */
import { DatabaseAdapter } from './adapter';
import { MySQLAdapter } from './mysql-adapter';
import { PostgresAdapter } from './postgres-adapter';
import { SQLiteAdapter } from './sqlite-adapter';

// Export all adapter modules
export * from './adapter';
export * from './base-adapter';
export * from './sqlite-adapter';
export * from './postgres-adapter';
export * from './mysql-adapter';

// Adapter registry
const adapters: Map<string, DatabaseAdapter> = new Map();

/**
 * Register a database adapter in the adapter registry
 */
export function registerAdapter(adapter: DatabaseAdapter): void {
	adapters.set(adapter.type, adapter);
}

/**
 * Get a database adapter by type
 * @param adapterType The type of adapter to get ('sqlite', 'mysql', 'postgres')
 * @returns The database adapter instance
 * @throws Error if adapter type is not found
 */
export function getAdapter(adapterType: string): DatabaseAdapter {
	const adapter = adapters.get(adapterType);
	if (!adapter) {
		throw new Error(`Database adapter '${adapterType}' not found`);
	}
	return adapter;
}

/**
 * Get all registered database adapters
 * @returns A record of all registered adapters, keyed by type
 */
export function getAdapters(): Record<string, DatabaseAdapter> {
	return Object.fromEntries(adapters);
}

// Register default adapters
registerAdapter(new SQLiteAdapter());
registerAdapter(new PostgresAdapter());
registerAdapter(new MySQLAdapter());
