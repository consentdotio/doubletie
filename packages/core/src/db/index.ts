/**
 * Database module exports
 * @module db
 */

// Core DB utilities
export { mergeSchemaWithConfig } from './merge';
export { generateTable } from './table';

// Database adapter exports
export {
	type DatabaseAdapter,
	BaseAdapter,
	SQLiteAdapter,
	PostgresAdapter,
	MySQLAdapter,
	// Adapter registry methods
	getAdapter,
	registerAdapter,
	getAdapters,
} from './adapters';

// Utility functions for working with databases
export {
	generateTableDefinition,
	generateSQLForEntity,
	toDatabaseValue,
	fromDatabaseValue,
} from './utils';

// Column and table definition types
export type {
	ColumnDefinition,
	TableDefinition,
	IndexDefinition,
	ForeignKeyDefinition,
} from './adapters/adapter';

// Import EntityFieldsMap from entity.types
export type { EntityFieldsMap } from '../entity/entity.types';

// Schema merging types
export type { ResolvedEntitySchema } from './merge';

/**
 * Database adapter type for type-safe adapter selection
 */
export type AdapterType = 'sqlite' | 'postgres' | 'mysql' | string;

/**
 * Helper function to generate a table definition for an entity
 * @param entity The entity definition
 * @param adapterType The database adapter type to use
 * @returns A table definition
 */
function generateTableDefinition<
	TEntity extends { name: string; fields: Record<string, any> },
>(entity: TEntity, adapterType: AdapterType = 'sqlite') {
	const { getAdapter } = require('./adapters');
	const adapter = getAdapter(adapterType);
	return adapter.generateTableDefinition(entity);
}

/**
 * Helper function to generate SQL for creating a table from an entity
 * @param entity The entity definition
 * @param adapterType The database adapter type to use
 * @returns SQL string for creating the table
 */
function generateSQLForEntity<
	TEntity extends { name: string; fields: Record<string, any> },
>(entity: TEntity, adapterType: AdapterType = 'sqlite'): string {
	const { getAdapter } = require('./adapters');
	const adapter = getAdapter(adapterType);
	const tableDef = adapter.generateTableDefinition(entity);
	return adapter.generateCreateTableSQL(tableDef);
}
