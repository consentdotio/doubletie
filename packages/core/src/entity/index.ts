/**
 * Entity module exports
 * @module entity
 */

// Entity definition exports
export { defineEntity } from './entity';
export { createRelationshipHelpers } from './relationship';

// Entity types
export type {
	EntityFields,
	EntityStructure,
	EntityFromDefinition,
	EntityWithRelationships,
	EntityInput,
	EntityOutput,
	EntityFieldsByType,
	WithRelationships,
} from './entity.types';

// Relationship types
export type {
	RelationshipType,
	RelationshipConfig,
	EntityFieldReference,
	RelationshipHelpers,
	BasicRelationshipOptions,
	ManyToManyOptions,
} from './relationship.types';
