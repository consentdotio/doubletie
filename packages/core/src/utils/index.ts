/**
 * Utility module exports
 * @module utils
 */

// Export type guards
export * from './type-guards';
