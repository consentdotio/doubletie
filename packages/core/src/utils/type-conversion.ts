import type { StandardSchemaV1 } from '@standard-schema/spec';
import type { EntityFromDefinition } from '../entity/entity.types';
import type {
	EntitySchemaDefinition,
	FieldValueType,
} from '../schema/schema.types';

/**
 * Convert EntityFromDefinition to EntitySchemaDefinition
 * This is needed for compatibility with database adapters
 */
export function toEntitySchemaDefinition<T extends EntityFromDefinition>(
	entity: T
): EntitySchemaDefinition {
	return {
		name: entity.name,
		prefix: entity.prefix,
		fields: entity.fields,
		config: entity.config as Record<string, FieldValueType>,
		order: entity.order,
		validator: entity.validator as StandardSchemaV1 | undefined,
		description: entity.description as string | undefined,
	};
}
