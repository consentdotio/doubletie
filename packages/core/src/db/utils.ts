/**
 * Database utility functions
 */
import { StandardSchemaV1 } from '@standard-schema/spec';
import type {
	EntitySchemaDefinition,
	FieldValueType,
	SchemaField,
} from '../schema/schema.types';
import { getAdapter } from './adapters';
import type { TableDefinition } from './adapters/adapter';

/**
 * Helper function to generate a table definition for an entity
 * @param entity The entity definition
 * @param adapterType The database adapter type to use
 * @returns A table definition
 */
export function generateTableDefinition(
	entity: EntitySchemaDefinition,
	adapterType = 'sqlite'
): TableDefinition {
	const adapter = getAdapter(adapterType);
	if (!adapter) {
		throw new Error(`Adapter not found for type: ${adapterType}`);
	}
	return adapter.generateTableDefinition(entity);
}

/**
 * Helper function to generate SQL for creating a table from an entity
 * @param entity The entity definition
 * @param adapterType The database adapter type to use
 * @returns SQL string for creating the table
 */
export function generateSQLForEntity(
	entity: EntitySchemaDefinition,
	adapterType = 'sqlite'
): string {
	const adapter = getAdapter(adapterType);
	const tableDef = adapter.generateTableDefinition(entity);
	return adapter.generateCreateTableSQL(tableDef);
}

/**
 * Helper function to transform a value to database format
 * @param field The field definition
 * @param value The value to transform
 * @param adapterType The database adapter type to use
 * @returns The transformed value
 */
export function toDatabaseValue<TField>(
	field: TField,
	value: SchemaField<
		string,
		FieldValueType,
		StandardSchemaV1<unknown, unknown>
	>,
	adapterType = 'sqlite'
): unknown {
	const adapter = getAdapter(adapterType);
	return adapter.toDatabase(field, value);
}
/**
 * Helper function to transform a value from database format
 * @param field The field definition
 * @param dbValue The database value to transform
 * @param adapterType The database adapter type to use
 * @returns The transformed value
 */
export function fromDatabaseValue<TField>(
	field: TField,
	dbValue: SchemaField<
		string,
		FieldValueType,
		StandardSchemaV1<unknown, unknown>
	>,
	adapterType = 'sqlite'
): any {
	const adapter = getAdapter(adapterType);
	return adapter.fromDatabase(field, dbValue);
}
