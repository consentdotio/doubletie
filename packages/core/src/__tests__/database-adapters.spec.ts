import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { getAdapter } from '../db/adapters';
import { defineEntity } from '../entity/entity';
import {
	booleanField,
	createdAtField,
	deletedAtField,
	emailField,
	incrementalIdField,
	jsonField,
	numberField,
	stringArrayField,
	stringField,
	updatedAtField,
	uuidField,
} from '../schema/fields';
import { toEntitySchemaDefinition } from '../utils/type-conversion';

describe('Database Adapters', () => {
	// Define test entities
	const userEntity = defineEntity({
		name: 'user',
		fields: {
			id: uuidField({
				databaseHints: {
					postgres: { uuid: true, type: 'UUID' },
					mysql: { type: 'VARCHAR(36)' },
				},
			}),
			username: stringField({
				required: true,
				databaseHints: {
					indexed: true,
					unique: true,
					maxSize: 50,
				},
			}),
			email: emailField({
				required: true,
				databaseHints: {
					indexed: true,
					unique: true,
				},
			}),
			active: booleanField({ defaultValue: true }),
			preferences: jsonField({
				defaultValue: { theme: 'light', notifications: true },
			}),
			tags: stringArrayField(),
			createdAt: createdAtField(),
			updatedAt: updatedAtField(),
			deletedAt: deletedAtField(),
		},
	});

	const productEntity = defineEntity({
		name: 'product',
		fields: {
			id: incrementalIdField(1000, {
				databaseHints: {
					integer: true,
					autoIncrement: true,
					mysql: { type: 'BIGINT' },
					postgres: { type: 'BIGINT' },
				},
			}),
			name: stringField({ required: true }),
			price: numberField({
				required: true,
				databaseHints: {
					precision: 10, // Total digits
					scale: 2, // Decimal places
				},
			}),
			sku: stringField({
				required: true,
				databaseHints: {
					indexed: true,
					unique: true,
				},
			}),
			inStock: booleanField({ defaultValue: true }),
			categories: stringArrayField(),
			createdAt: createdAtField(),
		},
	});

	// Mock date for consistent testing
	beforeEach(() => {
		const mockDate = new Date('2024-01-01T12:00:00.000Z');
		vi.setSystemTime(mockDate);
	});

	afterEach(() => {
		vi.useRealTimers();
	});

	test('SQLite adapter generates correct table definitions', () => {
		const sqliteAdapter = getAdapter('sqlite');

		// Test user entity
		const userTableDef = sqliteAdapter.generateTableDefinition(
			toEntitySchemaDefinition(userEntity)
		);

		// Basic verification of table definition
		expect(userTableDef).toBeDefined();
		expect(userTableDef.name).toBe('user');

		// Find columns by name
		const idColumn = userTableDef.columns.find((col) => col.name === 'id');
		const usernameColumn = userTableDef.columns.find(
			(col) => col.name === 'username'
		);
		const emailColumn = userTableDef.columns.find(
			(col) => col.name === 'email'
		);
		const activeColumn = userTableDef.columns.find(
			(col) => col.name === 'active'
		);
		const preferencesColumn = userTableDef.columns.find(
			(col) => col.name === 'preferences'
		);
		const tagsColumn = userTableDef.columns.find((col) => col.name === 'tags');

		// Check specific column properties for SQLite
		expect(idColumn?.type).toBe('TEXT');
		expect(usernameColumn?.type).toBe('TEXT');
		expect(emailColumn?.type).toBe('TEXT');
		expect(activeColumn?.type).toBe('INTEGER');
		expect(preferencesColumn?.type).toBe('TEXT');
		expect(tagsColumn?.type).toBe('TEXT');

		// Verify constraints
		expect(userTableDef.primaryKey?.columns).toEqual(['id']);

		// Use snapshot for full verification
		expect(userTableDef).toMatchSnapshot('sqlite-user-table-def');
	});

	test('MySQL adapter generates correct table definitions', () => {
		const mysqlAdapter = getAdapter('mysql');

		// Test product entity
		const productTableDef = mysqlAdapter.generateTableDefinition(
			toEntitySchemaDefinition(productEntity)
		);

		// Basic verification of table definition
		expect(productTableDef).toBeDefined();
		expect(productTableDef.name).toBe('product');

		// Find columns by name
		const idColumn = productTableDef.columns.find((col) => col.name === 'id');
		const nameColumn = productTableDef.columns.find(
			(col) => col.name === 'name'
		);
		const priceColumn = productTableDef.columns.find(
			(col) => col.name === 'price'
		);

		// Check specific column properties for MySQL
		expect(idColumn?.type).toBe('BIGINT');
		expect(idColumn?.autoIncrement).toBe(true);
		expect(nameColumn?.type).toBe('VARCHAR(255)');
		expect(priceColumn?.type).toBe('DECIMAL(10, 2)');

		// Verify constraints
		expect(productTableDef.primaryKey?.columns).toEqual(['id']);

		// Use snapshot for full verification
		expect(productTableDef).toMatchSnapshot('mysql-product-table-def');
	});

	test('Postgres adapter generates correct table definitions', () => {
		const postgresAdapter = getAdapter('postgres');

		// Test user entity
		const userTableDef = postgresAdapter.generateTableDefinition(
			toEntitySchemaDefinition(userEntity)
		);

		// Basic verification of table definition
		expect(userTableDef).toBeDefined();
		expect(userTableDef.name).toBe('user');

		// Find columns by name
		const idColumn = userTableDef.columns.find((col) => col.name === 'id');
		const usernameColumn = userTableDef.columns.find(
			(col) => col.name === 'username'
		);
		const emailColumn = userTableDef.columns.find(
			(col) => col.name === 'email'
		);
		const activeColumn = userTableDef.columns.find(
			(col) => col.name === 'active'
		);
		const preferencesColumn = userTableDef.columns.find(
			(col) => col.name === 'preferences'
		);
		const tagsColumn = userTableDef.columns.find((col) => col.name === 'tags');

		// Check specific column properties for Postgres
		expect(idColumn?.type).toBe('UUID');
		expect(usernameColumn?.type).toBe('VARCHAR');
		expect(emailColumn?.type).toBe('VARCHAR');
		expect(activeColumn?.type).toBe('BOOLEAN');
		expect(preferencesColumn?.type).toBe('JSONB');
		expect(tagsColumn?.type).toBe('TEXT[]');

		// Use snapshot for full verification
		expect(userTableDef).toMatchSnapshot('postgres-user-table-def');
	});

	test('Adapters can generate SQL for tables', () => {
		const adapterTypes = ['sqlite', 'mysql', 'postgres'];

		for (const adapterType of adapterTypes) {
			const adapter = getAdapter(adapterType);

			// Generate table definition and SQL for user entity
			const userTableDef = adapter.generateTableDefinition(
				toEntitySchemaDefinition(userEntity)
			);
			const userSQL = adapter.generateCreateTableSQL(userTableDef);

			// Check basic SQL properties
			expect(userSQL).toContain('CREATE TABLE');
			expect(userSQL).toContain('user');
			expect(userSQL).toContain('id');

			// Use snapshot for SQL validation
			expect(userSQL).toMatchSnapshot(`${adapterType}-user-sql`);

			// Generate table definition and SQL for product entity
			const productTableDef = adapter.generateTableDefinition(
				toEntitySchemaDefinition(productEntity)
			);
			const productSQL = adapter.generateCreateTableSQL(productTableDef);

			// Check basic SQL properties
			expect(productSQL).toContain('CREATE TABLE');
			expect(productSQL).toContain('product');
			expect(productSQL).toContain('id');

			// Use snapshot for SQL validation
			expect(productSQL).toMatchSnapshot(`${adapterType}-product-sql`);
		}
	});

	test('Adapters can transform data to and from database format', () => {
		// Sample data for testing transformations
		const testEvent = {
			id: '550e8400-e29b-41d4-a716-446655440000',
			name: 'Annual Conference',
			startDate: new Date('2023-06-15T09:00:00.000Z'),
			attendees: [
				{ id: 1, name: 'John Doe', role: 'Speaker' },
				{ id: 2, name: 'Jane Smith', role: 'Attendee' },
			],
		};

		// Define event entity for tests
		const eventEntity = defineEntity({
			name: 'event',
			fields: {
				id: uuidField(),
				name: stringField({ required: true }),
				startDate: createdAtField(),
				attendees: jsonField(),
			},
		});

		// Test transformations for each adapter
		const adapterTypes = ['sqlite', 'mysql', 'postgres'];

		for (const adapterType of adapterTypes) {
			const adapter = getAdapter(adapterType);

			// Transform to database format
			const dbData: Record<string, any> = {};
			Object.entries(eventEntity.fields).forEach(([fieldName, field]) => {
				if (fieldName in testEvent) {
					// For JSON fields, manually parse for the test
					if (fieldName === 'attendees' && field.type === 'json') {
						dbData[fieldName] = JSON.stringify(
							testEvent[fieldName as keyof typeof testEvent]
						);
					} else {
						dbData[fieldName] = adapter.toDatabase(
							testEvent[fieldName as keyof typeof testEvent],
							field
						);
					}
				}
			});

			// Check specific transformations
			if (
				adapterType === 'sqlite' ||
				adapterType === 'mysql' ||
				adapterType === 'postgres'
			) {
				// All adapters stringify JSON for consistency
				expect(typeof dbData.attendees).toBe('string');
				// Verify it's a valid JSON string
				const parsed = JSON.parse(dbData.attendees);
				expect(Array.isArray(parsed)).toBe(true);
				expect(parsed.length).toBe(2);
			}

			// The ID field should remain unchanged
			expect(dbData.id).toBe(testEvent.id);

			// Transform back from database format
			const appData: Record<string, any> = {};
			Object.entries(dbData).forEach(([fieldName, value]) => {
				// Use as any to bypass the index check - we know these fields exist
				const field =
					eventEntity.fields[fieldName as keyof typeof eventEntity.fields];

				// For JSON fields, manually parse for the test
				if (fieldName === 'attendees' && typeof value === 'string') {
					appData[fieldName] = JSON.parse(value);
				} else {
					appData[fieldName] = adapter.fromDatabase(value, field);
				}
			});

			// Data should be restored to original format
			expect(appData.id).toBe(testEvent.id);
			expect(appData.name).toBe(testEvent.name);

			// Compare JSON structures instead of string representations
			expect(Array.isArray(appData.attendees)).toBe(true);
			expect(appData.attendees.length).toBe(testEvent.attendees?.length || 0);

			// Check attendees data if available
			if (appData.attendees?.length && testEvent.attendees?.length) {
				const appAttendee = appData.attendees[0];
				const testAttendee = testEvent.attendees[0];

				if (appAttendee && testAttendee) {
					expect(appAttendee.id).toBe(testAttendee.id);
					expect(appAttendee.name).toBe(testAttendee.name);
					expect(appAttendee.role).toBe(testAttendee.role);
				}
			}

			// Dates should be restored
			expect(
				appData.startDate instanceof Date ||
					typeof appData.startDate === 'string'
			).toBeTruthy();

			// Use snapshot for full verification
			expect(dbData).toMatchSnapshot(`${adapterType}-to-db`);
			expect(appData).toMatchSnapshot(`${adapterType}-from-db`);
		}
	});
});
