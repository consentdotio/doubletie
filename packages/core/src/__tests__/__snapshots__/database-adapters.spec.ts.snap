// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Database Adapters > Adapters can generate SQL for tables > mysql-product-sql 1`] = `
"CREATE TABLE IF NOT EXISTS \`product\` (
  \`id\` BIGINT PRIMARY KEY AUTO_INCREMENT NOT NULL UNIQUE DEFAULT null,
  \`name\` VARCHAR(255) NOT NULL,
  \`price\` DECIMAL(10, 2) NOT NULL,
  \`sku\` VARCHAR(255) NOT NULL UNIQUE,
  \`inStock\` TINYINT(1) NULL DEFAULT true,
  \`categories\` TEXT NULL DEFAULT ,
  \`createdAt\` TIMESTAMP NOT NULL DEFAULT null,
  PRIMARY KEY (\`id\`)
);"
`;

exports[`Database Adapters > Adapters can generate SQL for tables > mysql-user-sql 1`] = `
"CREATE TABLE IF NOT EXISTS \`user\` (
  \`id\` VARCHAR(36) PRIMARY KEY NOT NULL UNIQUE DEFAULT null,
  \`username\` VARCHAR(255) NOT NULL UNIQUE,
  \`email\` VARCHAR(255) NOT NULL UNIQUE,
  \`active\` TINYINT(1) NULL DEFAULT true,
  \`preferences\` JSON NULL DEFAULT [object Object],
  \`tags\` TEXT NULL DEFAULT ,
  \`createdAt\` TIMESTAMP NOT NULL DEFAULT null,
  \`updatedAt\` TIMESTAMP NOT NULL DEFAULT null,
  \`deletedAt\` TIMESTAMP NULL,
  PRIMARY KEY (\`id\`)
);"
`;

exports[`Database Adapters > Adapters can generate SQL for tables > postgres-product-sql 1`] = `
"CREATE TABLE IF NOT EXISTS "product" (
  "id" BIGINT NOT NULL PRIMARY KEY UNIQUE DEFAULT NULL,
  "name" VARCHAR NOT NULL,
  "price" REAL NOT NULL,
  "sku" VARCHAR NOT NULL UNIQUE,
  "inStock" BOOLEAN DEFAULT 'TRUE',
  "categories" TEXT[] DEFAULT '''[]''::jsonb',
  "createdAt" TIMESTAMP NOT NULL DEFAULT NULL
);
"
`;

exports[`Database Adapters > Adapters can generate SQL for tables > postgres-user-sql 1`] = `
"CREATE TABLE IF NOT EXISTS "user" (
  "id" UUID NOT NULL PRIMARY KEY UNIQUE DEFAULT gen_random_uuid(),
  "username" VARCHAR NOT NULL UNIQUE,
  "email" VARCHAR NOT NULL UNIQUE,
  "active" BOOLEAN DEFAULT 'TRUE',
  "preferences" JSONB DEFAULT '''{"theme":"light","notifications":true}''::jsonb',
  "tags" TEXT[] DEFAULT '''[]''::jsonb',
  "createdAt" TIMESTAMP NOT NULL DEFAULT NULL,
  "updatedAt" TIMESTAMP NOT NULL DEFAULT NULL,
  "deletedAt" TIMESTAMP
);
"
`;

exports[`Database Adapters > Adapters can generate SQL for tables > sqlite-product-sql 1`] = `
"CREATE TABLE IF NOT EXISTS "product" (
  "id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE DEFAULT NULL,
  "name" TEXT NOT NULL DEFAULT NULL,
  "price" REAL NOT NULL DEFAULT NULL,
  "sku" TEXT NOT NULL UNIQUE DEFAULT NULL,
  "inStock" INTEGER DEFAULT '1',
  "categories" TEXT DEFAULT '''[]''',
  "createdAt" TEXT NOT NULL DEFAULT NULL
);"
`;

exports[`Database Adapters > Adapters can generate SQL for tables > sqlite-user-sql 1`] = `
"CREATE TABLE IF NOT EXISTS "user" (
  "id" TEXT PRIMARY KEY NOT NULL UNIQUE DEFAULT NULL,
  "username" TEXT NOT NULL UNIQUE DEFAULT NULL,
  "email" TEXT NOT NULL UNIQUE DEFAULT NULL,
  "active" INTEGER DEFAULT '1',
  "preferences" TEXT DEFAULT '''{"theme":"light","notifications":true}''',
  "tags" TEXT DEFAULT '''[]''',
  "createdAt" TEXT NOT NULL DEFAULT NULL,
  "updatedAt" TEXT NOT NULL DEFAULT NULL,
  "deletedAt" TEXT DEFAULT NULL
);"
`;

exports[`Database Adapters > Adapters can transform data to and from database format > mysql-from-db 1`] = `
{
  "attendees": [
    {
      "id": 1,
      "name": "John Doe",
      "role": "Speaker",
    },
    {
      "id": 2,
      "name": "Jane Smith",
      "role": "Attendee",
    },
  ],
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "name": "Annual Conference",
  "startDate": 2023-06-15T09:00:00.000Z,
}
`;

exports[`Database Adapters > Adapters can transform data to and from database format > mysql-to-db 1`] = `
{
  "attendees": "[{"id":1,"name":"John Doe","role":"Speaker"},{"id":2,"name":"Jane Smith","role":"Attendee"}]",
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "name": "Annual Conference",
  "startDate": 2023-06-15T09:00:00.000Z,
}
`;

exports[`Database Adapters > Adapters can transform data to and from database format > postgres-from-db 1`] = `
{
  "attendees": [
    {
      "id": 1,
      "name": "John Doe",
      "role": "Speaker",
    },
    {
      "id": 2,
      "name": "Jane Smith",
      "role": "Attendee",
    },
  ],
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "name": "Annual Conference",
  "startDate": 2023-06-15T09:00:00.000Z,
}
`;

exports[`Database Adapters > Adapters can transform data to and from database format > postgres-to-db 1`] = `
{
  "attendees": "[{"id":1,"name":"John Doe","role":"Speaker"},{"id":2,"name":"Jane Smith","role":"Attendee"}]",
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "name": "Annual Conference",
  "startDate": "2023-06-15T09:00:00.000Z",
}
`;

exports[`Database Adapters > Adapters can transform data to and from database format > sqlite-from-db 1`] = `
{
  "attendees": [
    {
      "id": 1,
      "name": "John Doe",
      "role": "Speaker",
    },
    {
      "id": 2,
      "name": "Jane Smith",
      "role": "Attendee",
    },
  ],
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "name": "Annual Conference",
  "startDate": 2023-06-15T09:00:00.000Z,
}
`;

exports[`Database Adapters > Adapters can transform data to and from database format > sqlite-to-db 1`] = `
{
  "attendees": "[{"id":1,"name":"John Doe","role":"Speaker"},{"id":2,"name":"Jane Smith","role":"Attendee"}]",
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "name": "Annual Conference",
  "startDate": "2023-06-15T09:00:00.000Z",
}
`;

exports[`Database Adapters > MySQL adapter generates correct table definitions > mysql-product-table-def 1`] = `
{
  "columns": [
    {
      "autoIncrement": true,
      "defaultValue": null,
      "name": "id",
      "nullable": false,
      "primaryKey": true,
      "type": "BIGINT",
      "unique": true,
    },
    {
      "autoIncrement": false,
      "name": "name",
      "nullable": false,
      "type": "VARCHAR(255)",
    },
    {
      "autoIncrement": false,
      "name": "price",
      "nullable": false,
      "type": "DECIMAL(10, 2)",
    },
    {
      "autoIncrement": false,
      "name": "sku",
      "nullable": false,
      "type": "VARCHAR(255)",
      "unique": true,
    },
    {
      "autoIncrement": false,
      "defaultValue": true,
      "name": "inStock",
      "nullable": true,
      "type": "TINYINT(1)",
    },
    {
      "autoIncrement": false,
      "defaultValue": [],
      "name": "categories",
      "nullable": true,
      "type": "TEXT",
    },
    {
      "autoIncrement": false,
      "defaultValue": null,
      "name": "createdAt",
      "nullable": false,
      "type": "TIMESTAMP",
    },
  ],
  "foreignKeys": [],
  "indexes": [],
  "name": "product",
  "options": {},
  "primaryKey": {
    "columns": [
      "id",
    ],
    "name": "pk_product",
  },
}
`;

exports[`Database Adapters > Postgres adapter generates correct table definitions > postgres-user-table-def 1`] = `
{
  "columns": [
    {
      "defaultValue": "gen_random_uuid()",
      "name": "id",
      "nullable": false,
      "primaryKey": true,
      "type": "UUID",
      "unique": true,
    },
    {
      "name": "username",
      "nullable": false,
      "type": "VARCHAR",
      "unique": true,
    },
    {
      "name": "email",
      "nullable": false,
      "type": "VARCHAR",
      "unique": true,
    },
    {
      "defaultValue": "TRUE",
      "name": "active",
      "nullable": true,
      "type": "BOOLEAN",
    },
    {
      "defaultValue": "'{"theme":"light","notifications":true}'::jsonb",
      "name": "preferences",
      "nullable": true,
      "type": "JSONB",
    },
    {
      "defaultValue": "'[]'::jsonb",
      "name": "tags",
      "nullable": true,
      "type": "TEXT[]",
    },
    {
      "defaultValue": null,
      "name": "createdAt",
      "nullable": false,
      "type": "TIMESTAMP",
    },
    {
      "defaultValue": null,
      "name": "updatedAt",
      "nullable": false,
      "type": "TIMESTAMP",
    },
    {
      "name": "deletedAt",
      "nullable": true,
      "type": "TIMESTAMP",
    },
  ],
  "foreignKeys": [],
  "indexes": [],
  "name": "user",
  "options": {},
  "primaryKey": {
    "columns": [
      "id",
    ],
    "name": "pk_user",
  },
}
`;

exports[`Database Adapters > SQLite adapter generates correct table definitions > sqlite-user-table-def 1`] = `
{
  "columns": [
    {
      "defaultValue": null,
      "name": "id",
      "nullable": false,
      "primaryKey": true,
      "type": "TEXT",
      "unique": true,
    },
    {
      "defaultValue": null,
      "name": "username",
      "nullable": false,
      "primaryKey": false,
      "type": "TEXT",
      "unique": true,
    },
    {
      "defaultValue": null,
      "name": "email",
      "nullable": false,
      "primaryKey": false,
      "type": "TEXT",
      "unique": true,
    },
    {
      "defaultValue": "1",
      "name": "active",
      "nullable": true,
      "primaryKey": false,
      "type": "INTEGER",
      "unique": false,
    },
    {
      "defaultValue": "'{"theme":"light","notifications":true}'",
      "name": "preferences",
      "nullable": true,
      "primaryKey": false,
      "type": "TEXT",
      "unique": false,
    },
    {
      "defaultValue": "'[]'",
      "name": "tags",
      "nullable": true,
      "primaryKey": false,
      "type": "TEXT",
      "unique": false,
    },
    {
      "defaultValue": null,
      "name": "createdAt",
      "nullable": false,
      "primaryKey": false,
      "type": "TEXT",
      "unique": false,
    },
    {
      "defaultValue": null,
      "name": "updatedAt",
      "nullable": false,
      "primaryKey": false,
      "type": "TEXT",
      "unique": false,
    },
    {
      "defaultValue": null,
      "name": "deletedAt",
      "nullable": true,
      "primaryKey": false,
      "type": "TEXT",
      "unique": false,
    },
  ],
  "foreignKeys": [],
  "indexes": [],
  "name": "user",
  "options": {},
  "primaryKey": {
    "columns": [
      "id",
    ],
    "name": "pk_user",
  },
}
`;
